gobblin-hive-registration/src/main/java/gobblin/hive/HiveRegister.java

```diff
@@ -0,0 +1,413 @@
+  }
+
+  private void createDbIfNotExists(HiveMetaStoreClient client, String dbName) {
+    Database db = new Database();
+    db.setName(dbName);
+
+    Preconditions.checkState(this.hiveDbRootDir.isPresent(),
+        "Missing required property " + HiveRegProps.HIVE_DB_ROOT_DIR);
+    db.setLocationUri(new Path(this.hiveDbRootDir.get(), dbName + HIVE_DB_EXTENSION).toString());
+
+    HiveLock lock = locks.getDbLock(dbName);
+    lock.lock();
+
+    try {
+      client.createDatabase(db);
+      log.info("Created database " + dbName);
+    } catch (AlreadyExistsException e) {
+      // Database already exists. Nothing to do.
+    } catch (TException e) {
+      log.error("Unable to create Hive database " + dbName);
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table createOrAlterTable(HiveMetaStoreClient client, String dbName, String tableName,
+      Optional<HivePartition> partition, StorageDescriptor sd) {
+
+    Table newTable = createTableWithoutRegistering(dbName, tableName, partition, sd);
+
+    HiveLock lock = locks.getTableLock(dbName, tableName);
+    lock.lock();
+
+    try {
+      if (client.tableExists(dbName, tableName)) {
+        Table existingTable = client.getTable(dbName, tableName);
+        if (needToUpdateTable(existingTable, newTable)) {
+          client.alter_table(dbName, tableName, newTable);
+          log.info(String.format("updated Hive table %s in db %s", tableName, dbName));
+          return newTable;
+        } else {
+          log.info(String.format("Hive table %s in db %s exists and no need to update", tableName, dbName));
+          return existingTable;
+        }
+      } else {
+        client.createTable(newTable);
+        log.info(String.format("Created Hive table %s in db %s", tableName, dbName));
+        return newTable;
+      }
+    } catch (TException e) {
+      log.error(String.format("Error in creating or altering Hive table %s in db %s", tableName, dbName));
+      throw Throwables.propagate(e);
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Table createTableWithoutRegistering(String dbName, String tableName, Optional<HivePartition> partition,
+      StorageDescriptor sd) {
+    Map<String, String> parameters = Maps.newHashMap();
+    parameters.put("EXTERNAL", Boolean.TRUE.toString());
+    Table table = new Table();
+    table.setParameters(parameters);
+    table.setDbName(dbName);
+    table.setTableName(tableName);
+
+    Preconditions.checkArgument(this.hiveOwner.isPresent(), "Missing required property " + HiveRegProps.HIVE_OWNER);
+    table.setOwner(this.hiveOwner.get());
+    table.setTableType(TableType.EXTERNAL_TABLE.toString());
+
+    if (partition.isPresent()) {
+      table.setPartitionKeys(partition.get().getKeys());
+    }
+
+    table.setSd(sd);
+    return table;
+  }
+
+  private boolean needToUpdateTable(Table existingTable, Table newTable) {
+    return !existingTable.equals(newTable);
```

Comment:
How is this going to work? Looking at http://grepcode.com/file/repo1.maven.org/maven2/org.apache.hive/hive-metastore/1.0.1/org/apache/hadoop/hive/metastore/api/Table.java#Table.equals%28org.apache.hadoop.hive.metastore.api.Table%29 a `Table` is compared based on its create time, so I would guess the result of `createTableWithoutRegistering` will never be equal to `existingTable`